src/sign/openssl.rs

```diff
@@ -1,58 +1,373 @@
+                let n = BigNum::from_slice(k.n.as_ref()).unwrap();
+                let e = BigNum::from_slice(k.e.as_ref()).unwrap();
+                let d = num(k.d.as_ref());
+                let p = num(k.p.as_ref());
+                let q = num(k.q.as_ref());
+                let d_p = num(k.d_p.as_ref());
+                let d_q = num(k.d_q.as_ref());
+                let q_i = num(k.q_i.as_ref());
+
+                // NOTE: The 'openssl' crate doesn't seem to expose
+                // 'EVP_PKEY_fromdata', which could be used to replace the
+                // deprecated methods called here.
+
+                openssl::rsa::Rsa::from_private_components(
+                    n, e, d, p, q, d_p, d_q, q_i,
+                )
+                .and_then(PKey::from_rsa)
+                .unwrap()
+            }
+            generic::SecretKey::EcdsaP256Sha256(k) => {
+                // Calculate the public key manually.
+                let ctx = openssl::bn::BigNumContext::new_secure().unwrap();
+                let group = openssl::nid::Nid::X9_62_PRIME256V1;
+                let group =
+                    openssl::ec::EcGroup::from_curve_name(group).unwrap();
+                let mut p = openssl::ec::EcPoint::new(&group).unwrap();
+                let n = num(k.as_slice());
+                p.mul_generator(&group, &n, &ctx).unwrap();
+                openssl::ec::EcKey::from_private_components(&group, &n, &p)
+                    .and_then(PKey::from_ec_key)
+                    .unwrap()
+            }
+            generic::SecretKey::EcdsaP384Sha384(k) => {
+                // Calculate the public key manually.
+                let ctx = openssl::bn::BigNumContext::new_secure().unwrap();
+                let group = openssl::nid::Nid::SECP384R1;
+                let group =
+                    openssl::ec::EcGroup::from_curve_name(group).unwrap();
+                let mut p = openssl::ec::EcPoint::new(&group).unwrap();
+                let n = num(k.as_slice());
+                p.mul_generator(&group, &n, &ctx).unwrap();
+                openssl::ec::EcKey::from_private_components(&group, &n, &p)
+                    .and_then(PKey::from_ec_key)
+                    .unwrap()
+            }
+            generic::SecretKey::Ed25519(k) => {
+                PKey::private_key_from_raw_bytes(
+                    k.as_ref(),
+                    pkey::Id::ED25519,
+                )
+                .unwrap()
+            }
+            generic::SecretKey::Ed448(k) => {
+                PKey::private_key_from_raw_bytes(k.as_ref(), pkey::Id::ED448)
+                    .unwrap()
+            }
+        };
+
+        Ok(Self {
+            algorithm: key.algorithm(),
+            pkey,
+        })
+    }
+
+    /// Export this key into a generic secret key.
+    ///
+    /// # Panics
+    ///
+    /// Panics if OpenSSL fails or if memory could not be allocated.
+    pub fn export<B>(&self) -> generic::SecretKey<B>
```

Comment:
I tend to think of import and export as referring to something outside the application, e.g. a file. I would suggest naming this "as_generic()`, or could this perhaps be written as a From impl?

Reply:
Yeah, I'll turn this into a `From` impl.  The primary reason to "export" to a generic key is indeed to serialize it on disk, but the additional layer of indirection makes the name confusing.